# Минипроект №1.
В первом мини-проекте вам предстоит проанализировать данные о **_бронировании отелей_**. Чуть ниже вы можете ознакомиться со списком вопросов и описанием данных. Для удобства проверки, далее задачи разбиты на стэпы.
### Задачи
* Импортируйте библиотеку pandas как pd. Загрузите датасет bookings.csv с разделителем ;. Проверьте размер таблицы, типы переменных, а затем выведите первые 7 строк, чтобы посмотреть на данные.

* Приведите названия колонок к нижнему регистру и замените пробелы на знак нижнего подчеркивания.

* Пользователи из каких стран совершили наибольшее число успешных бронирований? Укажите топ-5.

* На сколько ночей в среднем бронируют отели разных типов?

* Иногда тип номера, полученного клиентом (assigned_room_type), отличается от изначально забронированного (reserved_room_type).Такое может произойти, например, по причине овербукинга. Сколько подобных наблюдений встретилось в датасете?

* Проанализируйте даты запланированного прибытия. – На какой месяц чаще всего успешно оформляли бронь в 2016? Изменился ли самый популярный месяц в 2017?– Сгруппируйте данные по годам и проверьте, на какой месяц бронирования отеля типа City Hotel отменялись чаще всего в каждый из периодов.

* Посмотрите на числовые характеристики трёх переменных: adults, children и babies. Какая из них имеет наибольшее среднее значение?

* Создайте колонку total_kids, объединив children и babies. Для отелей какого типа среднее значение переменной оказалось наибольшим?


### Задача №1. 
Импортируйте библиотеку pandas как pd. Прочитайте датасет bookings.csv с разделителем ;. Ссылка на датасет находится в переменной path_to_file.

Сохраните прочитанный датасет в bookings.

Затем запишите первые 7 строк в переменную bookings_head.

In [1]:
import pandas as pd
import warnings
warnings.filterwarnings("ignore")

In [2]:
bookings = pd.read_csv('data/bookings.csv', encoding = 'windows-1251', sep = ';')
bookings.head(10)

Hotel  Is Canceled  Lead Time arrival full date  Arrival Date Year  \
0  Resort Hotel            0        342        2015-07-01               2015   
1  Resort Hotel            0        737        2015-07-01               2015   
2  Resort Hotel            0          7        2015-07-01               2015   
3  Resort Hotel            0         13        2015-07-01               2015   
4  Resort Hotel            0         14        2015-07-01               2015   
5  Resort Hotel            0         14        2015-07-01               2015   
6  Resort Hotel            0          0        2015-07-01               2015   
7  Resort Hotel            0          9        2015-07-01               2015   
8  Resort Hotel            1         85        2015-07-01               2015   
9  Resort Hotel            1         75        2015-07-01               2015   

  Arrival Date Month  Arrival Date Week Number  Arrival Date Day of Month  \
0               July                        27                          1   
1               July                        27                          1   
2               July                        27                          1   
3               July                        27                          1   
4               July                        27                          1   
5               July                        27                          1   
6               July                        27                          1   
7               July                        27                          1   
8               July                        27                          1   
9               July                        27                          1   

   Stays in Weekend nights  Stays in week nights  ...  Adults  Children  \
0                        0                     0  ...       2       0.0   
1                        0                     0  ...       2       0.0   
2                        0                     1  ...       1       0.0   
3                        0                     1  ...       1       0.0   
4                        0                     2  ...       2       0.0   
5                        0                     2  ...       2       0.0   
6                        0                     2  ...       2       0.0   
7                        0                     2  ...       2       0.0   
8                        0                     3  ...       2       0.0   
9                        0                     3  ...       2       0.0   

   Babies  Meal Country Reserved Room Type Assigned room type customer type  \
0       0    BB     PRT                  C                  C     Transient   
1       0    BB     PRT                  C                  C     Transient   
2       0    BB     GBR                  A                  C     Transient   
3       0    BB     GBR                  A                  A     Transient   
4       0    BB     GBR                  A                  A     Transient   
5       0    BB     GBR                  A                  A     Transient   
6       0    BB     PRT                  C                  C     Transient   
7       0    FB     PRT                  C                  C     Transient   
8       0    BB     PRT                  A                  A     Transient   
9       0    HB     PRT                  D                  D     Transient   

  Reservation Status Reservation status_date  
0          Check-Out              2015-07-01  
1          Check-Out              2015-07-01  
2          Check-Out              2015-07-02  
3          Check-Out              2015-07-02  
4          Check-Out              2015-07-03  
5          Check-Out              2015-07-03  
6          Check-Out              2015-07-03  
7          Check-Out              2015-07-03  
8           Canceled              2015-05-06  
9           Canceled              2015-04-22  

[10 rows x 21 columns]

In [3]:
bookings_head = bookings.head(7) 
bookings_head

Hotel  Is Canceled  Lead Time arrival full date  Arrival Date Year  \
0  Resort Hotel            0        342        2015-07-01               2015   
1  Resort Hotel            0        737        2015-07-01               2015   
2  Resort Hotel            0          7        2015-07-01               2015   
3  Resort Hotel            0         13        2015-07-01               2015   
4  Resort Hotel            0         14        2015-07-01               2015   
5  Resort Hotel            0         14        2015-07-01               2015   
6  Resort Hotel            0          0        2015-07-01               2015   

  Arrival Date Month  Arrival Date Week Number  Arrival Date Day of Month  \
0               July                        27                          1   
1               July                        27                          1   
2               July                        27                          1   
3               July                        27                          1   
4               July                        27                          1   
5               July                        27                          1   
6               July                        27                          1   

   Stays in Weekend nights  Stays in week nights  ...  Adults  Children  \
0                        0                     0  ...       2       0.0   
1                        0                     0  ...       2       0.0   
2                        0                     1  ...       1       0.0   
3                        0                     1  ...       1       0.0   
4                        0                     2  ...       2       0.0   
5                        0                     2  ...       2       0.0   
6                        0                     2  ...       2       0.0   

   Babies  Meal Country Reserved Room Type Assigned room type customer type  \
0       0    BB     PRT                  C                  C     Transient   
1       0    BB     PRT                  C                  C     Transient   
2       0    BB     GBR                  A                  C     Transient   
3       0    BB     GBR                  A                  A     Transient   
4       0    BB     GBR                  A                  A     Transient   
5       0    BB     GBR                  A                  A     Transient   
6       0    BB     PRT                  C                  C     Transient   

  Reservation Status Reservation status_date  
0          Check-Out              2015-07-01  
1          Check-Out              2015-07-01  
2          Check-Out              2015-07-02  
3          Check-Out              2015-07-02  
4          Check-Out              2015-07-03  
5          Check-Out              2015-07-03  
6          Check-Out              2015-07-03  

[7 rows x 21 columns]

### Задача №2.
Посмотрите на размеры таблицы. Сколько всего в ней столбцов?

In [4]:
"Количество столбцов = {}".format(bookings.shape[1])

'Количество столбцов = 21'

### Задача №3.
К какому типу относится большинство переменных? 
* object
* int64
* float64

In [5]:
bookings.dtypes 

Hotel                         object
Is Canceled                    int64
Lead Time                      int64
arrival full date             object
Arrival Date Year              int64
Arrival Date Month            object
Arrival Date Week Number       int64
Arrival Date Day of Month      int64
Stays in Weekend nights        int64
Stays in week nights           int64
stays total nights             int64
Adults                         int64
Children                     float64
Babies                         int64
Meal                          object
Country                       object
Reserved Room Type            object
Assigned room type            object
customer type                 object
Reservation Status            object
Reservation status_date       object
dtype: object

Ответ: int64, object 

### Задача №4.
Как вы уже могли заметить, при сборе датасета произошла небольшая ошибка, и названия колонок были записаны в неудобном для дальнейшей работы формате. Исправьте ситуацию, приведя названия столбцов к нижнему регистру и заменив пробелы на знак нижнего подчеркивания. 
**Например: Is Canceled -> is_canceled**

In [6]:
# Функция, котороя приводит названия столбцов к нижнему регистру и меняет пробел между словами на _.
def adsc(name):
    new_name=name.lower().replace(' ', '_') 
    return new_name                         

In [7]:
bookings = bookings.rename(columns=adsc) 
bookings.head(5)

hotel  is_canceled  lead_time arrival_full_date  arrival_date_year  \
0  Resort Hotel            0        342        2015-07-01               2015   
1  Resort Hotel            0        737        2015-07-01               2015   
2  Resort Hotel            0          7        2015-07-01               2015   
3  Resort Hotel            0         13        2015-07-01               2015   
4  Resort Hotel            0         14        2015-07-01               2015   

  arrival_date_month  arrival_date_week_number  arrival_date_day_of_month  \
0               July                        27                          1   
1               July                        27                          1   
2               July                        27                          1   
3               July                        27                          1   
4               July                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  ...  adults  children  \
0                        0                     0  ...       2       0.0   
1                        0                     0  ...       2       0.0   
2                        0                     1  ...       1       0.0   
3                        0                     1  ...       1       0.0   
4                        0                     2  ...       2       0.0   

   babies  meal country reserved_room_type assigned_room_type customer_type  \
0       0    BB     PRT                  C                  C     Transient   
1       0    BB     PRT                  C                  C     Transient   
2       0    BB     GBR                  A                  C     Transient   
3       0    BB     GBR                  A                  A     Transient   
4       0    BB     GBR                  A                  A     Transient   

  reservation_status reservation_status_date  
0          Check-Out              2015-07-01  
1          Check-Out              2015-07-01  
2          Check-Out              2015-07-02  
3          Check-Out              2015-07-02  
4          Check-Out              2015-07-03  

[5 rows x 21 columns]

### Задача №5.
Пользователи из каких стран совершили наибольшее число успешных бронирований? Бронирование считается успешным, если в дальнейшем не было отменено (переменная is_canceled). В качестве ответа выберите страны, входящие в **топ-5**.

In [8]:
bookings.query("reservation_status == 'Check-Out'").groupby('country', as_index = False) \
        .agg({'reservation_status':'count'}) \
        .sort_values('reservation_status', ascending=False).head(5)

country  reservation_status
125     PRT               21071
57      GBR                9676
54      FRA                8481
50      ESP                6391
42      DEU                6069

## Задача №6.
 На сколько ночей (stays_total_nights)  в среднем бронируют отели типа **City Hotel**? **Resort Hotel**? Запишите полученные значения в пропуски с точностью до 2 знаков после точки.

In [9]:
bookings.groupby('hotel', as_index = False) \
          .agg({'stays_total_nights':'mean'}).rename(columns={'stays_total_nights':'mean_total_nights'}) \
          .sort_values('mean_total_nights', ascending=False).round(2) 

hotel  mean_total_nights
1  Resort Hotel               4.32
0    City Hotel               2.98

### Задача №7.
Иногда тип номера, присвоенного клиенту (assigned_room_type), отличается от изначально забронированного (reserved_room_type). Такое может произойти, например, по причине овербукинга.

Сколько подобных наблюдений встретилось в датасете?Отмена бронирования также считается

In [10]:
f"Ответ: {bookings.query('assigned_room_type != reserved_room_type').shape[0]}"

'Ответ: 14917'

### Задача №8.
Теперь проанализируйте даты запланированного прибытия (arrival_date_year).

На какой месяц чаще всего оформляли бронь в 2016 году? Изменился ли самый популярный месяц в 2017?

1. В 2017 чаще всего бронировали отель на октябрь
2. В 2016 чаще всего бронировали отель на май
3. В 2016 чаще всего оформляли бронирование на сентябрь
4. В 2016 чаще всего бронировали отель на октябрь
5. В 2017 чаще всего оформляли бронирование на май

In [11]:
bookings.groupby(['arrival_date_month', 'arrival_date_year'], as_index=False) \
        .agg({'reservation_status_date' : 'count'}) \
        .sort_values('arrival_date_month', ascending=False) \
        .rename(columns = {'reservation_status_date':'count_reservation_by_date'})

arrival_date_month  arrival_date_year  count_reservation_by_date
25          September               2016                       5394
24          September               2015                       5114
23            October               2016                       6203
22            October               2015                       4957
21           November               2016                       4454
20           November               2015                       2340
19                May               2017                       6313
18                May               2016                       5478
17              March               2017                       4970
16              March               2016                       4824
14               June               2016                       5292
15               June               2017                       5647
13               July               2017                       5313
12               July               2016                       4572
11               July               2015                       2776
10            January               2017                       3681
9             January               2016                       2248
8            February               2017                       4177
7            February               2016                       3891
6            December               2016                       3860
5            December               2015                       2920
4              August               2017                       4925
3              August               2016                       5063
2              August               2015                       3889
1               April               2017                       5661
0               April               2016                       5428

Ответ:

4. В 2016 чаще всего бронировали отель на октябрь

5. В 2017 чаще всего оформляли бронирование на май


### Задача №9.
Сгруппируйте данные по годам, а затем проверьте, на какой месяц (arrival_date_month) бронирования отеля типа **City Hotel** отменялись чаще всего в 2015? 2016? 2017? 

In [12]:
bookings.query("is_canceled == 1 and hotel == 'City Hotel'") \
        .groupby(['arrival_date_month', 'arrival_date_year'], as_index=False) \
        .agg({'reservation_status' : 'count'}) \
        .sort_values(['arrival_date_year', 'reservation_status'], ascending=False)

arrival_date_month  arrival_date_year  reservation_status
19                May               2017                2217
1               April               2017                1926
15               June               2017                1808
13               July               2017                1324
17              March               2017                1278
4              August               2017                1123
10            January               2017                1044
8            February               2017                 971
23            October               2016                1947
14               June               2016                1720
25          September               2016                1567
0               April               2016                1539
18                May               2016                1436
21           November               2016                1360
3              August               2016                1247
16              March               2016                1108
6            December               2016                1072
12               July               2016                1043
7            February               2016                 930
9             January               2016                 438
24          September               2015                1543
22            October               2015                1321
2              August               2015                1232
11               July               2015                 939
5            December               2015                 668
20           November               2015                 301

Ответ:

_Май_ - 2017

_Октябрь_ - 2016

_Сентябрь_ - 2015

### Задача №10.
Посмотрите на числовые характеристики трёх колонок: **adults**, **children** и **babies**. Какая из них имеет наибольшее среднее значение?

In [13]:
bookings[{'adults', 'children', 'babies'}].describe()

children         adults         babies
count  119386.000000  119390.000000  119390.000000
mean        0.103890       1.856403       0.007949
std         0.398561       0.579261       0.097436
min         0.000000       0.000000       0.000000
25%         0.000000       2.000000       0.000000
50%         0.000000       2.000000       0.000000
75%         0.000000       2.000000       0.000000
max        10.000000      55.000000      10.000000

Ответ : **abults** с средним значением **1.856403**

### Задача №11.
Создайте колонку total_kids, объединив столбцы children и babies. Для отелей какого типа среднее значение переменной оказалось наибольшим?

* City hotel – отель находится в городе
* Resort hotel – отель курортный

В качестве ответа укажите наибольшее среднее total_kids, округлив до 2 знаков после точки

In [14]:
bookings['total_kids'] = bookings.children + bookings.babies

In [15]:
bookings.groupby('hotel', as_index=False) \
        .agg({'total_kids' : 'mean'}) \
        .sort_values('total_kids', ascending=False) \
        .rename(columns = {'total_kids':'mean_total_kids'}).round(2)

hotel  mean_total_kids
1  Resort Hotel             0.14
0    City Hotel             0.10

**Ответ :** наибольшее среднее значение переменной total_kids равно 0.14 для Resort Hotel